src/ir_def/construct.ml

```diff
@@ -89,7 +89,11 @@ let primE prim es =
     | ICCallerPrim -> T.caller
     | ICStableRead t -> t
     | ICStableWrite _ -> T.unit
-    | IdxPrim -> T.(as_immut (as_array (List.hd es).note.Note.typ))
+    | IdxPrim
+    | DerefArrayOffset -> T.(as_immut (as_array (List.hd es).note.Note.typ))
+    | NextArrayOffset _ -> T.nat
+    | ValidArrayOffset -> T.bool
+    | GetPastArrayOffset _ -> T.nat
```

Comment:
The return type of GetPastArrayOffset _ is a lie?

Reply:
It is not a lie while in the scope of IR. _After_ leaving IR (entering Wasm, most types being erased), not a problem any more. I have described this in the PR commentary.

Reply:
There is not a single IR typing lie. The new primitives all play by the typing rules.